File: app/src/main/java/rht/raytracer/Camera.java
package rht.raytracer;

public class Camera {
    private final Vec3 focalPoint;
    private final Vec3 frameCentre;
    private final Vec3 xDirection;
    private final Vec3 yDirection;

    public Camera(Vec3 focalPoint, Vec3 frameCentre, Vec3 xDirection, Vec3 yDirection) {
        this.focalPoint = focalPoint;
        this.frameCentre = frameCentre;
        this.xDirection = xDirection;
        this.yDirection = yDirection;
    }

    public Ray rayForPixel(double frameX, double frameY) {
        Vec3 framePoint = frameCentre.plus(xDirection.times(frameX)).plus(yDirection.times(frameY));
        Vec3 direction = framePoint.minus(focalPoint);
        return new Ray(focalPoint, direction);
    }
}

File: rustray/src/camera.rs
use crate::{maths::vec::Vector, ray::Ray};

#[derive(Clone, Debug, PartialEq)]
pub struct Camera {
    pub focal_point: Vector<3>,
    pub frame_centre: Vector<3>,
    pub x_direction: Vector<3>,
    pub y_direction: Vector<3>,
}

impl Camera {
    pub fn ray_for_pixel(&self, frame_x: f64, frame_y: f64) -> Ray {
        let frame_point =
            &(&self.frame_centre + &(&self.x_direction * frame_x)) + &(&self.y_direction * frame_y);
        let direction = &frame_point - &self.focal_point;
        Ray::new(&self.focal_point, &direction)
    }
}
